gnome-image-installer/pages/confirm/gis-confirm-page.c

```diff
@@ -0,0 +1,281 @@
+
+#include <udisks/udisks.h>
+#include <glib/gstdio.h>
+#include <glib/gi18n.h>
+#include <gio/gio.h>
+#include <stdlib.h>
+#include <errno.h>
+
+struct _GisConfirmPagePrivate {
+    GtkLabel *vendor_label;
+    GtkLabel *product_label;
+
+    GtkLabel *image1_label;
+    GtkLabel *model1_label;
+    GtkLabel *device1_label;
+    GtkLabel *size1_label;
+
+    GtkBox *warning_box;
+    GtkLabel *warning_label;
+
+    GtkBox *wrong_computer_box;
+
+    GtkLabel *countdown_label;
+    GtkButton *cancel_button;
+    GtkButton *continue_button;
+
+    guint countdown_source;
+    guint countdown_remaining;
+};
+typedef struct _GisConfirmPagePrivate GisConfirmPagePrivate;
+
+G_DEFINE_TYPE_WITH_PRIVATE (GisConfirmPage, gis_confirm_page, GIS_TYPE_PAGE);
+G_DEFINE_QUARK(confirm-error, gis_confirm_error);
+
+static void
+gis_confirm_page_advance (GisConfirmPage *self)
+{
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+  GisPage *page = GIS_PAGE (self);
+
+  if (priv->countdown_source != 0)
+    {
+      g_source_remove (priv->countdown_source);
+      priv->countdown_source = 0;
+    }
+
+  gis_assistant_next_page (gis_driver_get_assistant (page->driver));
+}
+
+static gboolean
+gis_confirm_page_countdown_cb (gpointer data)
+{
+  GisConfirmPage *self = GIS_CONFIRM_PAGE (data);
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+
+  if (priv->countdown_remaining == 0)
+    {
+      gis_confirm_page_advance (self);
+      return G_SOURCE_REMOVE;
+    }
+  else
+    {
+      g_autofree gchar *countdown_message =
+        g_strdup_printf (_("Installation will commence automatically in %d seconds."),
+                         priv->countdown_remaining);
+      gtk_label_set_text (priv->countdown_label, countdown_message);
+      priv->countdown_remaining--;
+      return G_SOURCE_CONTINUE;
+    }
+}
+
+static void
+gis_confirm_page_start_countdown (GisConfirmPage *self)
+{
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+
+  priv->countdown_remaining = 30;
+  gis_confirm_page_countdown_cb (self);
+
+  priv->countdown_source = g_timeout_add_seconds (1,
+                                                  gis_confirm_page_countdown_cb,
+                                                  self);
+}
+
+static void
+gis_confirm_page_shown (GisPage *page)
+{
+  GisConfirmPage *self = GIS_CONFIRM_PAGE (page);
+  GisConfirmPagePrivate *priv = gis_confirm_page_get_instance_private (self);
+  GFile *image_file;
+  g_autofree gchar *image_basename = NULL;
+  UDisksClient *udisks_client;
+  UDisksBlock *target_block;
+  UDisksDrive *target_drive = NULL;
+  g_autofree gchar *target_model = NULL;
+  g_autofree gchar *target_size = NULL;
+  /* TODO */
+  gboolean target_computer_specified = TRUE;
+  gboolean target_computer_matches = TRUE;
+
+  if (gis_store_get_error() != NULL)
+    {
+      gis_confirm_page_advance (self);
+      return;
+    }
+
+  /* TODO: update DMI labels */
```

Comment:
(I assume this `TODO` is what the commit message refers to as being updated in a future commit. 👍 )